kai-analyzer/pkg/service/analyzer.go

```diff
@@ -22,84 +22,91 @@ type Analyzer struct {
 	cancelFunc context.CancelFunc
 
 	initedProviders map[string]provider.InternalProviderClient
-	ruleSets        []engine.RuleSet
 }
 
-func NewAnalyzer(limitIncidents, limitCodeSnips, contextLines int, location, incidentSelector string, ruleFiles []string, log logr.Logger) (*Analyzer, error) {
+func NewAnalyzer(log logr.Logger) *Analyzer {
+	return &Analyzer{
+		Logger:          log,
+		initedProviders: map[string]provider.InternalProviderClient{},
+	}
+}
+
+func (a *Analyzer) Initialize(args InitializeArgs) error {
 	ctx, cancelFunc := context.WithCancel(context.Background())
 	eng := engine.CreateRuleEngine(ctx,
-		10,
-		log,
-		engine.WithIncidentLimit(limitIncidents),
-		engine.WithCodeSnipLimit(limitCodeSnips),
-		engine.WithContextLines(contextLines),
-		engine.WithIncidentSelector(incidentSelector),
+		args.Workers,
+		a.Logger,
+		engine.WithIncidentLimit(args.LimitIncidents),
+		engine.WithCodeSnipLimit(args.LimitCodeSnips),
+		engine.WithContextLines(args.ContextLines),
+		engine.WithIncidentSelector(args.IncidentSelector),
 	)
 
-	// this function already init's the java provider
-	jProvider, err := java.NewInternalProviderClient(ctx, log, contextLines, location)
+	javaProvider := javaexp.NewJavaProvider(a.Logger, "java", args.ContextLines, provider.Config{Name: "java"})
+	_, _, err := javaProvider.Init(ctx, a.Logger, provider.InitConfig{
+		Location: args.Location,
+		ProviderSpecificConfig: map[string]interface{}{
+			"lspServerName": "java",
+			"bundles":       args.JavaConfig.Bundles,
+			"lspServerPath": args.JavaConfig.LSPServerPath,
+		},
+		Proxy:        &provider.Proxy{},
+		AnalysisMode: provider.AnalysisMode(args.AnalysisMode),
+	})
 	if err != nil {
 		cancelFunc()
-		return nil, err
+		return err
 	}
+	_, err = javaProvider.ProviderInit(ctx, []provider.InitConfig{})
+	if err != nil {
+		cancelFunc()
+		return err
+	}
+	a.initedProviders["java"] = javaProvider
 
-	bProvider, err := lib.GetProviderClient(provider.Config{Name: "builtin"}, log)
+	builtinProvider, err := lib.GetProviderClient(provider.Config{Name: "builtin"}, a.Logger)
 	if err != nil {
 		cancelFunc()
-		return nil, err
+		return err
 	}
-	_, err = bProvider.ProviderInit(context.Background(), []provider.InitConfig{{Location: location}})
+	_, _, err = builtinProvider.Init(ctx, a.Logger, provider.InitConfig{})
 	if err != nil {
 		cancelFunc()
-		return nil, err
+		return err
 	}
-
-	providers := map[string]provider.InternalProviderClient{
-		"java":    jProvider,
-		"builtin": bProvider,
+	_, err = builtinProvider.ProviderInit(context.Background(), []provider.InitConfig{{Location: args.Location}})
+	if err != nil {
+		cancelFunc()
+		return err
 	}
+	a.initedProviders["builtin"] = builtinProvider
+
+	a.engine = eng
```

Comment:
Are we ever going to have more than one initial requests?

Reply:
I imagine each project creating a new instance so no

Reply:
We should not, although we are not explicitly forbidding it yet. I modeled this exchange after LSP where, if you need to re`initialize` you would have to call `shutdown` and `exit` before calling `initialize` again.